segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,182 @@
+        val chunksToDelete = new ArrayList<String>();
+        int count = 0;
+        for (val entry : garbageChunks.entrySet()) {
+            if (canDelete(entry.getValue(), currentTime)) {
+                chunksToDelete.add(entry.getKey());
+                // Throttle how many chunks are deleted at once.
+                if (++count >= maxItems) {
+                    break;
+                }
+            }
+        }
+
+        ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (val chunkToDelete : chunksToDelete) {
+            val txn = chunkedSegmentStorage.getMetadataStore().beginTransaction(false, chunkToDelete);
+            val future =
+                    txn.get(chunkToDelete)
+                    .thenComposeAsync( metadata -> {
+                            val chunkMetadata = (ChunkMetadata) metadata;
+                            // Check whether the chunk is marked as inactive
+                            if (!chunkMetadata.isActive()) {
+                                return chunkedSegmentStorage.getChunkStorage().openWrite(chunkToDelete)
+                                        .thenComposeAsync(chunkedSegmentStorage.getChunkStorage()::delete, chunkedSegmentStorage.getExecutor())
+                                        .thenRunAsync(() -> {
+                                            garbageChunks.remove(chunkToDelete);
+                                            txn.delete(chunkToDelete);
+                                            log.debug("{} deleteGarbage - deleted chunk={}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                        }, chunkedSegmentStorage.getExecutor())
+                                        .thenComposeAsync(v -> txn.commit(), chunkedSegmentStorage.getExecutor())
+                                        .exceptionally(e -> {
+                                            val ex = Exceptions.unwrap(e);
+                                            if (ex instanceof ChunkNotFoundException) {
+                                                log.debug("{} deleteGarbage - Could not delete garbage chunk {}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                            } else {
+                                                log.warn("{} deleteGarbage - Could not delete garbage chunk {}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                            }
+                                            return null;
+                                        });
+                            } else {
+                                log.info("{} deleteGarbage - Chunk is not marked as garbage{}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                return CompletableFuture.completedFuture(null);
+                            }
+                        })
+                    .whenCompleteAsync((v, ex) -> {
```

Comment:
Which executor are you running this on? Most likely it will go onto `ForkJoinPool.commonPool()` if you don't explicitly pass one.

Reply:
fixed.